lib/notificationHandler.js

```diff
@@ -0,0 +1,297 @@
+
+    /**
+     * Create a new instance of NotificationHandler
+     *
+     * @param {object} settings - settings like states/objects db
+     */
+    constructor(settings) {
+        this.states = settings.states;
+        this.objects = settings.objects;
+        this.log = settings.log;
+        this.currentNotifications = {};
+        this.setup = {};
+        // default data dir is relative to main.js
+        this.dataDir = path.join(__dirname, '..', tools.getDefaultDataDir());
+        this.logPrefix = settings.logPrefix;
+        this.host = settings.host;
+
+        // load notifications file
+        this._loadNotifications();
+    }
+
+    /**
+     * Get all adapter instances on this host and store their notifications config
+     *
+     * @return {Promise<void>}
+     */
+    async getSetupOfAllAdaptersFromHost() {
+        try {
+            const res = await this.objects.getObjectViewAsync('system', 'instance', {
+                startkey: 'system.adapter.',
+                endkey: 'system.adapter.\u9999'
+            });
+
+            if (res && Array.isArray(res.rows)) {
+                for (const entry of res.rows) {
+                    // check that instance is running on given host and has notifications
+                    if (entry && entry.value && entry.value.notifications && entry.value.common && entry.value.common.host === this.host) {
+                        this.addConfig(entry.value.notifications, entry.id);
+                    }
+                }
+            }
+        } catch (e) {
+            throw new Error(`Could not get notifications setup from instances: ${e.message}`);
+        }
+
+    }
+
+    /**
+     * Add a new category to the given scope with provided optional list of regex
+     *
+     * @param {object[]} notifications - notifications array
+     * @param {string} instance - e.g. hm-rpc.1 or host
+     */
+    addConfig(notifications, instance) {
+        if (instance !== this.host && !instance.startsWith('system.adapter.')) {
+            instance = 'system.adapter.' + instance;
+        }
+
+        // if valid attributes, store it
+        if (Array.isArray(notifications)) {
+            for (const scopeObj of notifications) {
+                if (Array.isArray(scopeObj.categories)) {
+                    for (const categoryObj of scopeObj.categories) {
+                        this.setup[scopeObj.scope] = this.setup[scopeObj.scope] || {};
+                        this.setup[scopeObj.scope][categoryObj.category] = this.setup[scopeObj.scope][categoryObj.category] || {};
+                        try {
+                            // regex is an array
+                            const regex = categoryObj.regex;
+                            for (let i = 0; i < regex.length; i++) {
+                                regex[i] = new RegExp(categoryObj.regex[i]);
+                            }
+                            this.setup[scopeObj.scope][categoryObj.category] = {
```

Comment:
Important for documentation: a new overwritten config overwrites fully. We can add special handling for "merging" later if needed (e,g if akready existing just overwrite whats "new" :-) (regex or such)

Reply:
Yes, I've also added a code comment